### PR TITLE
Performance enhancement reducing clone

### PR DESCRIPTION
**Pull request by patefacio**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -29,7 +29,10 @@ url = "1"
 humansize = "1"
 # used in date format filter
 chrono = "0.4"
+# used in instrument timing
```
Comment (Keats) [inline Cargo.toml]: the 3 dependencies below should be in dev-dependencies then?

File: src/bin/big_context_no_loop.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
+extern crate tera;
```
Comment (Keats) [inline src/bin/big_context_no_loop.rs]: is this file needed? can it be in bench instead?

File: src/renderer/ast_processor.rs [hunk 3]
```diff
@@ -0,0 +1,1064 @@
+//! Responsible for walking ast and render implementation
+
+// --- module use statements ---
```
Comment (Keats) [inline src/renderer/ast_processor.rs]: can you remove all those comments?

File: src/renderer/tests/errors.rs [hunk 4]
```diff
@@ -109,12 +102,12 @@ fn error_unknown_index_variable() {
...
-    assert_eq!(
-        result.unwrap_err().iter().nth(1).unwrap().description(),
-        "Variable arr[a] can not be evaluated because: Variable `a` not found in context while rendering \'tpl\'"
-    );
+    let result_text = result_text(tera.render("tpl", &Context::new()));
+println!("ERROR: {}", result_text); // TODO
+    // assert_eq!(
+    //     result.unwrap_err().iter().nth(1).unwrap().description(),
+    //     "Variable arr[a] can not be evaluated because: Variable `a` not found in context while rendering \'tpl\'"
+    // );
```
Comment (Keats) [inline src/renderer/tests/errors.rs]: it should be uncommented no?

File: Cargo.toml [hunk 5]
```diff
@@ -29,7 +29,10 @@ url = "1"
...
+stopwatch = "^0.0.7"
+log = "^0.3.8"
```
Comment (Keats) [inline Cargo.toml]: was logging used only when debugging? can it be removed?

Comment (patefacio) [reply Cargo.toml]: Removed stopwatch dependency. I do use `log!` by creating a macro called `out!` that points to `log!` for dev but can switch to `println!` if desire is to see logs from testing, since testing does not support.

I can remove this logging if you'd like, but maybe best to leave until you have time to review, in case need to review code more generally and logging might help. Whichever you prefer.

File: src/renderer/tests/errors.rs (hunk 4 above)
Comment (patefacio) [reply src/renderer/tests/errors.rs]: Replaced this with a test.

File: src/renderer/ast_processor.rs (hunk 3 above)
Comment (patefacio) [reply src/renderer/ast_processor.rs]: Removed.

File: src/bin/big_context_no_loop.rs (hunk 2 above)
Comment (patefacio) [reply src/bin/big_context_no_loop.rs]: Removed file.

File: Cargo.toml (hunk 5 above)
Comment (patefacio) [reply Cargo.toml]: I believe I effectively did a rebase. My first and my first pull request, so I may have screwed it up. I hope it helps you diff more easily - but I'm not seeing how it helps so maybe I missed a step.

File: src/renderer/context.rs [hunk 6]
```diff
@@ -0,0 +1,150 @@
+use context::get_json_pointer;
+use serde_json::value::Value;
+
+// --- module struct definitions ---
```
Comment (Keats) [inline src/renderer/context.rs]: sorry I meant to remove these comments in every files

Comment (patefacio) [reply src/renderer/context.rs]: Ok, removed those.

**Comment by Keats** [comment]
Since the diff was so big, I ended up rewriting this PR to make sure I understood it well enough to maintain the codebase properly: the result is in https://github.com/Keats/tera/pull/327

A few comments:
- `RefOrOwned` exists in the std as [Cow](https://doc.rust-lang.org/stable/std/borrow/enum.Cow.html)
- it uncovered a few bugs with macros that I fixed but still need some digging/additional tests 
- the new structure is WAY cleaner than the old one: much easier to understand where we are not to mention allocating way less

I need to check if there are other places I can avoid allocating but overall thanks a lot for that PR, I marked you as author of #327  as it is mostly your work after all.

**Comment by patefacio** [comment]
This is great! Thanks. Someone recommended COW when I described the
use-case and when I read its docs it sounded very different than my needs.
So, it will be interesting to see how you used it.
